### PR TITLE
UInt128: Throw DivideByZeroException when divisor is 0

### LINKED ISSUE #94593: {U}Int128 divided by 0 throws wrong exception type
Repro:
```C#
Int128 a = Int128.MaxValue;
Int128 b = 0;
Int128 result = a / b;
```
or
```C#
UInt128 a = UInt128.MaxValue;
UInt128 b = 0;
UInt128 result = a / b;
```

If `a` is set to a small value, like 42, this results in the expected `DivideByZeroException`. But when `a` is set to a large value, like `MaxValue`, it instead throws an ArgumentOutOfRangeException.

```
Unhandled exception. System.ArgumentOutOfRangeException: Specified argument was out of the range of valid values.
   at System.UInt128.<op_Division>g__DivideSlow|110_2(UInt128 quotient, UInt128 divisor)
   at System.UInt128.op_Division(UInt128 left, UInt128 right)
   at System.Int128.op_Division(Int128 left, Int128 right)
```

### PR DESCRIPTION
**Pull request by manandre**
Fixes #94593

### CODE REVIEW

File: src/libraries/System.Private.CoreLib/src/System/UInt128.cs [hunk 1]
```diff
@@ -1091,6 +1091,11 @@ public static UInt128 Log2(UInt128 value)
         /// <inheritdoc cref="IDivisionOperators{TSelf, TOther, TResult}.op_Division(TSelf, TOther)" />
         public static UInt128 operator /(UInt128 left, UInt128 right)
         {
+            if ((right._lower == 0) && (right._upper == 0))
+            {
+                throw new DivideByZeroException();
```
Comment (danmoseley) [inline src/libraries/System.Private.CoreLib/src/System/UInt128.cs:1096]: Should this extract the throw info a helper method?

File: src/libraries/System.Private.CoreLib/src/System/UInt128.cs [hunk 2]
```diff
@@ -1091,6 +1091,11 @@ public static UInt128 Log2(UInt128 value)
...
+            }
+
             if ((right._upper == 0) && (left._upper == 0))
```
Comment (danmoseley) [inline src/libraries/System.Private.CoreLib/src/System/UInt128.cs:1099]: Is it better or worse to combine the check of right._upper == 0 with the one above?

Comment (skyoxZ) [reply src/libraries/System.Private.CoreLib/src/System/UInt128.cs:1099]: If we don't combine them, I guess putting the DevideByZero check after the both-int64 check could bring better performance.

File: src/libraries/System.Private.CoreLib/src/System/UInt128.cs (hunk 1 above)
Comment (manandre) [reply src/libraries/System.Private.CoreLib/src/System/UInt128.cs]: Yes, this is a good idea.

File: src/libraries/System.Private.CoreLib/src/System/UInt128.cs (hunk 2 above)
Comment (manandre) [reply src/libraries/System.Private.CoreLib/src/System/UInt128.cs]: Checks combined.
However, as a general rule, I prefer to keep input validation first, before routing the call to the `UInt64` divide operator.
@danmoseley Is the check order important here?

Comment (danmoseley) [reply src/libraries/System.Private.CoreLib/src/System/UInt128.cs]: Area owners should answer that.

Comment (tannergooding) [reply src/libraries/System.Private.CoreLib/src/System/UInt128.cs]: We should prefer throwing exceptions for "invalid" inputs first.

**Review by tannergooding — APPROVED** [review]

**Comment by tannergooding** [comment]
There is a conflict that needs resolving before this can be merged.

**Comment by manandre** [comment]
@tannergooding Commits rebased. The conflict was only about some test files moved.
See https://github.com/dotnet/runtime/pull/94644

**Comment by danmoseley** [comment]
thanks @manandre for the contribution!

**Pull request merged**